### DISCUSSION TITLE
Redirecting libmesh and PETSc stdout/stderr

### DISCUSSION THREAD

**Discussion by rebpoli** [libMesh/libmesh#3040] [Q&A] [answered]
Hi,

I am having trouble redirecting libmesh and PETSc stdout/stderr.
I tried a number of alternatives, but I could not get to a clean solution.

I tried this kind of construction:
        libMesh::out.rdbuf (ostream->rdbuf());
        libMesh::err.rdbuf (ostream->rdbuf());

But KSP messages from PETSc (from -ksp_monitor) are not forwarded.
Then I tried to redirect PETSc messages using PETSc's "petscVFPrint" function.
Little success so far... 

Any idea? Shouldn't petsc messages be forwarded to libMesh::out and ::err?

Thanks,
Renato

**Comment by jwpeterson** [comment, marked as answer]
To redirect libmesh output, you should pass  `--keep-cout --redirect-output` on the command line. This will write all output to files called "log.processor.N" by default, where N is the processor rank. Regarding PETSc output, I don't believe this will be affected by our redirecting of `std::cout` because their print statements are (at some level) based around using C `printf` statements.

**Comment by giorgiobornia** [comment]
As a solution, one may consider a mini-wrapper script where shell redirection " > " to a certain file is performed.
For sure, this will collect all outputs from all linked libraries, Petsc or whatever else.

If one wants this filename, or these filenames, to match with the libmesh ones, 
 then it is true that one needs to generate such names in the shell script, and pass them via command line.

 Basically, it is a bit of "externalization" from the program out to a script, but perhaps redirection in the end
 is more of a "shell" thing than anything.

Giorgio
